Check restaurants

newoo opened the pull request:
안녕하세요.

아직 구현을 반도 못했지만, 방향성 점검 겸 리뷰 요청합니다

newoo committed:
Add api test that fetch regions and categories

newoo committed:
Add api to pass test about fetch regions and categories

newoo committed:
Update api by async/await style

newoo committed:
Update api.js to use network

newoo committed:
Update api.test.jsx to test about fetching restaurants data

newoo committed:
Update api.jsx to pass test about fetching restaurants

newoo committed:
Add Buttons.test.jsx to test Buttons functions

wholemann reviewed (changes requested):
안녕하세요 태헌님~

일단 API 테스트는 백엔드 쪽에서 잘 됐음을 믿고

보통 프론트엔드에선 REST API 동작 자체에 대해선 테스트 하지 않습니다.

그리고 프론트엔드에선 상태에 대한 비지니스 로직 처리가 테스트 대상인데
사실 이건 REST API와는 상관이 없습니다.

앱 입장에선 상태를 REST API를 통해 가져오는지 아니면 브라우저 내부의
local storage와 같은 곳에서 가져오는지는 관심사가 아닙니다.

실제로 API가 작동하거나 제공되지 않아도 앱이 동작하는데 문제가 없어야 합니다.

그래서 아샬님의 강의를 보시면 실제로 등장하는 테스트는 reducer.test.js 임을 보실 수 있습니다.

TDD에서 제일 중요한 건 사용하는 측(테스트 코드)에서 사용법을 먼저 정의하고
그에 맞춰 실제 로직을 구현하는 겁니다.
이는 테스트 코드가 없어도 실천해야하는 습관입니다. 사용하는 측에서 먼저 어떻게 사용할지를 정하는 거죠.



지금은 왜 Buttons부터 등장했는진 모르겠네요.

저라면...
기본적인 App 껍데기만 만들고 이 부분에서 우리가 필요한 상태가 무엇인지 생각해 봐야죠.
과제 요구사항을 보면 어쨌든 초기 상태를 어딘가에서 가져와야 한다는 걸 알 수 있죠.
그러면 action에 대한 구현과 이를 사용하는 reducer 테스트가 먼저 나와야 할 겁니다.

다음 세 가지 규칙을 지켜보길 권합니다:)

1. 실패한 단위 테스트를 만들기 전에는 제품 코드를 만들지 않는다.
2. 컴파일이 안 되거나 실패한 단위 테스트가 있으면 더 이상의 단위 테스트를 만들지 않는다.
3. 실패한 단위 테스트를 통과하는 이상의 제품 코드는 만들지 않는다.

newoo committed:
Add Buttons.jsx to pass test

newoo committed:
Update index.jsx to apply store

newoo committed:
Add store.js

newoo committed:
Add actions.js

newoo committed:
Refactor api.js

newoo committed:
Add reducer.test.js to test initial state of reducer

newoo committed:
Add reducer.js to pass initial state test

newoo committed:
Update codes about reducer to pass choose region test

newoo committed:
Update codes about reducer to pass choose category test

newoo committed:
Add ButtonContainer.test.jsx to test dispatching region choice

newoo commented:
다시 한 번 방향성 체크 부탁드려요
action과 reducer구현을 기존 코드에서 추가해봤고, ButtonsContainer 테스트 및 기능을 추가 구현 중입니다.
Buttons를 먼저 구현해서 컨테이너를 추후 개발하게 되었는데,

**사용하는 측에서 먼저 어떻게 사용할지**를 고려한다면 컨테이너 테스트 코드 및 구현 코드를 먼저하는 것이 좋은 방향인가요??

다음 리뷰 요청 전에 다시 한 번 TDD 복습의 필요성을 느끼네요;

wholemann commented on src/reducer.test.js:
```diff
@@ -0,0 +1,40 @@
+  });
+
+  describe('chooseRegion', () => {
+    context('when click region button', () => {
```
reducer는 UI와는 전혀 관련이 없습니다. 여기서 click이 등장하면 안 됩니다.

reducer는 click이 되는지 어떤지 아예 몰라야 해요.

wholemann commented on src/actions.js:
```diff
@@ -0,0 +1,17 @@
+export function chooseRegion(regionName) {
```
상태는 고르는 게 아니기 때문에 set이나 change와 같은 동사로 보통 표현이 됩니다.

현재 표현은 UI에서 일어난 일을 표현하고 있습니다.

그리고 UI에서는 select가 오히려 맞는 표현 같네요.

wholemann commented on src/services/api.test.js:
```diff
@@ -0,0 +1,57 @@
+import { fetchRegions, fetchCategories, fetchRestaurants } from './api';
+
+describe('api', () => {
```
단위 테스트는 굉장히 자주 실행되는데, API 호출이 들어가면 일단 속도가 느려지는 게 첫번째 문제구요.

가장 문제는 외부에서 제공하는 API를 이렇게 할 경우 공격으로 인식될 가능성이 큽니다.

E2E테스트로 검증하지 프론트엔드 앱에서 API 자체를 이렇게 테스트하는 경우는 없다고 보셔야 합니다.

wholemann commented on src/ButtonsContainer.test.jsx:
```diff
@@ -0,0 +1,50 @@
+
+jest.mock('react-redux');
+
+describe('ButtonsContainer', () => {
```
Button은 아무런 의미를 나타내지 못하고 있습니다.

RegionsContainer 같은 게 맞죠. 표현이 Button일 뿐 그 디자인은 언제든 바뀔 수 있습니다.

디자인이 바뀌면 이름이 어떻게 될까요?

wholemann reviewed (changes requested):
네 그건 그때 그때 편한 식으로 하면 되는데

말씀하신대로 일단 Container부터 만들고 또 Container에서 작업하다 보니

분리할 컴포넌트가 나오면 해당 컴포넌트 테스트를 추가하면서 진행하겠죠.

몇 가지 아쉬운 부분 comment 남겨드렸습니다. 참고하셔요ㅎㅎ 질문은 DM 주세요!

newoo committed:
Update context text in reducer test

newoo committed:
Update name of actions from choose~ to set~

newoo committed:
Delete api.test.js

newoo committed:
Update file name from ButtonContainers.test to RegionsContainer.test

newoo committed:
Delete no need codes

newoo committed:
Add RegionsContainer to pass test about dispatching setRegion

newoo committed:
Add test codes about select regions

newoo committed:
Delete codes that do not need

newoo committed:
Add CategoriesContainer test about select categories and then add CategoriesContainer

newoo committed:
Add test about dispatching setCategory

test to dispatch setCategory but other test suddenly error, so update actions, reducer and Buttons

newoo committed:
Add test code about rendering regions and categories

newoo committed:
Update App to pass test

newoo committed:
Update store by adding code about redux-trunk

newoo committed:
Add codes to get regions data from api server

newoo committed:
Update codes to pass rendering regions and categories test again

wholemann commented on src/reducer.test.js:
```diff
@@ -17,7 +17,7 @@ describe('reducer', () => {
 
   describe('chooseRegion', () => {
-    context('when click region button', () => {
+    context('with a new region name', () => {
```
https://youtu.be/gfTsSBRvdqI

지금은 context가 등장할 이유가 없습니다. 맥락이 갈라지지 않기 때문이죠.

describe context it 에 대한 사용법은 위의 영상을 참고해주세요.

newoo committed:
Add actions about get category from api server

wholemann commented on src/RegionsContainer.test.jsx:
```diff
@@ -0,0 +1,39 @@
+
+jest.mock('react-redux');
+
+describe('RegionsContainer', () => {
```
지금은 catergory와 regions가 같이 있네요.

둘은 엄밀히 다른 상태를 가지고 있습니다.

관심사의 분리를 떠올려 보시면 좋겠네요.

wholemann commented on src/Buttons.jsx:
```diff
@@ -0,0 +1,13 @@
+import React from 'react';
+
+export default function Buttons({ values, onClick }) {
```
다소 성급한 추상화인 것 같습니다.

Buttons가 어떤 의미를 갖나요? 이름으론 굉장히 범용적으로 쓰이는

디자인 시스템에 있을 컴포넌트입니다. 과연 이 Buttons는 범용적으로 쓰일 수 있을까요?

모든 컴포넌트에서 같은 상태를 내려줄 수 있을까요?

만약 쓰이는 곳에서 내려주는 상태가 달라지면 어떻게 될까요?

wholemann reviewed (changes requested):
안녕하세요~

일단 전체적으로 아샬님 강의를 다시 보시면서 진행하시면 좋겠네요ㅎㅎ

몇 가지 아쉬운 부분 comment 드렸습니다.

제가 submit을 안 눌렀었네요ㅎㅎ;

질문은 언제든 DM으로 환영합니다.

newoo committed:
Update codes related with get category data from api server

newoo committed:
Add test code about rendering restaurants

newoo committed:
Add codes to pass rendering restaurants

newoo committed:
Add codes to get restaurants action

newoo committed:
Add RestaurantsContainer test to test selecting restaurants

newoo committed:
Add RestaurantsContainer to pass test

newoo commented:
오전 피드백을 못 보고 과제 진행을 했었네요;;

오전 리뷰에 대한 수정사항은 다음 작업에 추가할께요..

wholemann commented on src/CategoriesContainer.jsx:
```diff
@@ -13,8 +13,8 @@ export default function RegionsContainer() {
 
-  function handleClickChooseCategory(categoryId) {
-    dispatch(setCategory(categoryId));
+  function handleClickChooseCategory(category) {
```
지금은 handleClick 정도면 충분할 것 같네요.

wholemann commented on src/Restaurants.test.jsx:
```diff
@@ -0,0 +1,29 @@
+
+describe('Restaurants', () => {
+  context('with restaurants', () => {
+    const testRestaurants = [
```
굳이 testRestaurants를 하실 필욘 없어요.

그냥 사용하는 곳이랑 똑같이 restaurants로 선언하시면 됩니다.

테스트는 문서니까요.

wholemann reviewed (commented):
지금 수준의 복잡도에선 Restaurants로 까지 나눌 필욘 없어 보이긴 합니다ㅎㅎ

보통 저는 일단 RestaurantsContainer 랑 같은 파일 안에 

상단 부분에 Restaurants만 분리해놓고 파일까진 나누지 않다가

Restaurants의 책임이 무거워지기 시작하면 테스트를 추가하고 파일을 분리하거든요.

사실 처음부터 미리 나누실 필요는 없어요ㅎㅎ 물론 연습하는 걸 목표로 나눠 보는 것도 좋습니다만

적정 수준은 고민해볼 필요가 있습니다.

newoo committed:
Refactor code to be able to reuse in RegionsContainer.test

newoo committed:
Delete no need contexts

newoo committed:
Update funtions name in containers

newoo committed:
Update object name for test

newoo committed:
Add test for Regions

newoo committed:
Add Regions component

newoo committed:
Update all codes about Regions according to new components

newoo committed:
Add categories compoent test

newoo committed:
Add Categories component

newoo committed:
Update codes to fix infinite api calls

newoo committed:
Update Containers of Regions and Categories to pass coverage

newoo commented:
아직 커버리지 테스트 등을 통과하지 못한 상황이지만, 시간 관계상 우선 PR합니다.
통과 후에 다시 PR하는 방향으로 진행해볼께요

action과 api 테스트 커버리지의 경우, 어떻게 처리하는 것이 좋나요?(늦은 밤이라, DM 대신에 여기에 질문 남깁니다.)

newoo committed:
Update codes related with reducer to pass test coverage

wholemann commented on src/App.test.jsx:
```diff
@@ -0,0 +1,45 @@
+describe('App', () => {
+  context('with regions and categories', () => {
+    it('renders regions and categories', () => {
+      const testRegions = [
```
test는 굳이 안 붙이셔도 될 것 같아요.

이 테스트 코드에선 test가 아니라 실제 사용되는 값이기 때문이죠.

wholemann commented on src/Categories.jsx:
```diff
@@ -0,0 +1,24 @@
+        <li key={category.id}>
+          <button type="button" onClick={() => onClick(category)}>
+            {category.name}
+            {getClickedMarkIfClicked(category.id)}
```
삼항 연산자를 이용해서 이곳에서 바로 표현하면 더 좋을 것 같습니다.

일단 복잡도가 크지 않고 오히려 render가 어떻게 될지 그 '의도'가 바로 보이질 않네요.

wholemann commented on src/Categories.test.jsx:
```diff
@@ -0,0 +1,54 @@
+
+import Categories from './Categories';
+
+describe('Categories', () => {
```
보통 

context('with ~', () =>) 나오면

context('without ~', () =>) 쌍으로 나오는 상황이 보통입니다.

이 테스트에서 등장하는 context가 과연 전혀 다른 맥락을 표현하는지 생각해볼 필요가 있습니다.

wholemann commented on src/CategoriesContainer.jsx:
```diff
@@ -0,0 +1,30 @@
+
+  function handleClick(category) {
+    dispatch(setCategory(category));
+    dispatch(loadRestaurants(regionName, category.id));
```
loadRestaurants 에서 getState를 이용한다면 이 파라미터 전달은 불필요하죠.

wholemann commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,53 @@
+  });
+
+  context('with a new category id', () => {
+    it('dispatch setCategory', () => {
```
dispatch보단 listens click event가 오히려 주된 관심사 같습니다.

wholemann commented on src/CategoriesContainer.test.jsx:
```diff
@@ -0,0 +1,53 @@
+  ];
+
+  context('with categories', () => {
+    it('select categories', () => {
```
select가 아니라

it renders categories가 맞지 않을까요?

wholemann commented on src/actions.js:
```diff
@@ -0,0 +1,76 @@
+}
+
+export function loadCategories() {
+  return async (dispatch) => {
```
```suggestion
  return async (dispatch,  getState) => {
```

wholemann reviewed (commented):
api의 경우 jest.mockResolvedValue를 이용하여

mockFetch를 만들어서

URL과 상관없이 해당 함수 호출의 결과값 형태를 검증하는 방식으로 해결할 수 있습니다.

Slack에서도 논의됐던 내용이죠.

저는 실무에선 굳이 테스트하진 않습니다만ㅎㅎ API는 어련히 백엔드쪽에서 잘 만들었어야 하고

실제로 테스트한다면 E2E로 하는 게 확실하죠.

action도 마찬가지로 복잡한 로직이 없이 단순히 객체 리턴이 전부라 굳이 테스트하진 않습니다만

커버리지를 위해선 그냥 제대로 원하는 형태가 리턴됐는지만 점검하면 됩니다.

테스트 통과까지 계속 진행하시고 남겨드린 comment와 아샬님 강의 참고하셔서

꼭 완전한 형태로 PR 날리시고 Merge 받으시면 좋겠습니다!
